lib/src/test/java/io/ably/lib/test/realtime/RealtimeConnectFailTest.java

```diff
@@ -210,25 +202,88 @@ public Object getTokenRequest(TokenParams params) throws AblyException {
 			opts.logLevel = Log.VERBOSE;
 			AblyRealtime ably = new AblyRealtime(opts);
 
-            ably.connection.on(new ConnectionStateListener() {
-                @Override
-                public void onConnectionStateChanged(ConnectionStateChange state) {
-                    if (state.previous == ConnectionState.connected) {
-                        synchronized (flags) {
-                            flags[1] = true;
-                            flags.notify();
-                        }
-                    }
-                }
-            });
-
-            synchronized (flags) {
-                try {
-                    flags.wait(15000);
-                } catch (InterruptedException e) {}
-            }
-
-            assertTrue("Verify that token generation was called", flags[0]);
+			/* wait for connected state */
+			ConnectionWaiter connectionWaiter = new ConnectionWaiter(ably.connection);
+			connectionWaiter.waitFor(ConnectionState.connected);
+			assertEquals("Verify connected state is reached", ConnectionState.connected, ably.connection.state);
+
+			/* wait for disconnected state (on token expiry), with timeout */
+			connectionWaiter.waitFor(ConnectionState.disconnected, 1, 30000L);
+			assertEquals("Verify disconnected state is reached", ConnectionState.disconnected, ably.connection.state);
+
+			/* wait for connected state (on token renewal) */
+			connectionWaiter.waitFor(ConnectionState.connected, 1, 30000L);
+			assertEquals("Verify connected state is reached", ConnectionState.connected, ably.connection.state);
+
+			/* verify that our token generator was called */
+			assertEquals("Expected token generator to be called", 1, authCallback.getCbCount());
+
+			/* end */
+			ably.close();
+			connectionWaiter.waitFor(ConnectionState.closed);
+			assertEquals("Verify closed state is reached", ConnectionState.closed, ably.connection.state);
+		} catch (AblyException e) {
+			e.printStackTrace();
+			fail("init0: Unexpected exception instantiating library");
+		}
+	}
+
+	/**
+	 * Verify that the server issues reauth message 30 seconds before token expiration time, authCallback is
+	 * called to obtain new token and in-place re-authorization takes place with connection staying in connected
+	 * state
+	 */
+	@Test
+	public void connect_token_expire_inplace_reauth() {
+		try {
+			final Setup.TestVars optsTestVars = Setup.getTestVars();
+			ClientOptions optsForToken = optsTestVars.createOptions(optsTestVars.keys[0].keyStr);
+			optsForToken.logLevel = Log.VERBOSE;
+			final AblyRest ablyForToken = new AblyRest(optsForToken);
+			TokenDetails tokenDetails = ablyForToken.auth.requestToken(new TokenParams() {{ ttl = 34000L; }}, null);
```

Comment:
Why 34s, if we want to test what happens when the token expires, why not go for a short TTL?